plugins/guisupport/guisupport.cpp

```diff
@@ -499,29 +501,33 @@ QIcon GuiSupport::createIcon(const QIcon &oldIcon, QWindow *w)
         gammarayIcon.addFile(QLatin1String(":/gammaray/images/gammaray-inject-128.png"));
     }
 
-    const QIcon &orgIcon = m_originalIcons[w];
-    if (!oldIcon.isNull() && (!orgIcon.isNull() || orgIcon.cacheKey() == oldIcon.cacheKey())) {
+    // Make sure to override the window icon with the app icon if it was initialized with it
+    if (w) {
+        const auto it = m_originalIcons.constFind(qApp);
+        const QIcon appIcon = it != m_originalIcons.constEnd() ? it.value() : qApp->windowIcon();
+        if (oldIcon.cacheKey() == appIcon.cacheKey() || oldIcon.cacheKey() == qApp->windowIcon().cacheKey()) {
+            return qApp->windowIcon();
+        }
+    }
+
+    const auto it = m_originalIcons.constFind(w ? (QObject *)w : (QObject *)qApp);
+    if (!oldIcon.isNull() && (it != m_originalIcons.constEnd() && (!it.value().isNull() || it.value().cacheKey() == oldIcon.cacheKey()))) {
```

Comment:
we now can directly use it != m_originalIcons.constEnd() istead of isNull.
isNull was used to test whether the default constructor was used.